components/places/sql/create_shared_triggers.sql

```diff
@@ -4,13 +4,20 @@
 
 -- This file defines triggers shared between the main and Sync connections.
 
+-- markh doesn't understand why this trigger is needed. It seems to be doing roughly
+-- the same work as moz_updateoriginsinsert_afterdelete_trigger and so we have 2 very
+-- similar triggers for the same insert. A difference between them though is some
+-- frecency stuff - and if you remove this filter, all tests pass except one obscure
+-- one in matcher.rs, where a test site ends up with a frecency of zero rather than the
+-- expected 1999. Both triggers exist in desktop too.
+-- So if anyone can explain this, please update this comment :)
```

Comment:
@mhammond I figured it out—the [order of the deletes](https://github.com/mozilla/application-services/blob/6c4c65e8ce5808edd12260195e222ed988d780e5/components/places/src/storage/mod.rs#L384-L388) in `delete_pending_temp_tables` is off! 🤯

I think the order of those deletes has to be:

```
DELETE FROM moz_updateoriginsinsert_temp;
DELETE FROM moz_updateoriginsupdate_temp;
DELETE FROM moz_updateoriginsdelete_temp;
```

Specifically, it's important that `DELETE FROM moz_updateoriginsinsert_temp` runs _before_ `DELETE FROM moz_updateoriginsupdate_temp`. Here's what I think is happening, and where the bug is:

When `apply_observation_direct` adds visits for a never-before-visited URL, it does three things:

1. Inserts a row into `moz_places` for the new URL.
2. Inserts rows into `moz_historyvisits` for the new visits.
3. Updates `moz_places.frecency` for the new URL to account for the visits.

(1) fires the `moz_places_afterinsert_trigger_origins` trigger, and (3) fires the `moz_places_afterupdate_frecency_trigger`.

Currently, `delete_pending_temp_tables` does [`DELETE FROM moz_updateoriginsupdate_temp` first](https://github.com/mozilla/application-services/blob/6c4c65e8ce5808edd12260195e222ed988d780e5/components/places/src/storage/mod.rs#L385). That'll fire the `moz_updateoriginsupdate_afterdelete_trigger`, which will try to update `moz_origins.frecency` for the new URL's origin.

_But_, there isn't a row for the URL in `moz_origins` yet—because that row is inserted by the `moz_updateoriginsinsert_afterdelete_trigger`, and that trigger won't fire until we do `DELETE FROM moz_updateoriginsinsert_temp`!

So [this `UPDATE moz_origins` statement](https://github.com/mozilla/application-services/blob/6c4c65e8ce5808edd12260195e222ed988d780e5/components/places/sql/create_shared_triggers.sql#L226-L229) won't do anything, and the initial frecency for the new origin—2000, same as Desktop—will be silently dropped.

When `DELETE FROM moz_updateoriginsinsert_temp` eventually runs, it'll insert a row for the URL's origin (via the `moz_updateoriginsinsert_afterdelete_trigger`), but with an initial origin frecency of 0.

[Desktop's behavior is correct here](https://searchfox.org/mozilla-central/rev/9ac4c7d32fa21ddef5cdf281939228ecfedabf64/toolkit/components/places/History.cpp#991-1009): inserts for new origins are processed before frecency updates for those new origins.

With the order of the deletes fixed, the frecency also changes to match Desktop's! 🎉

TL;DR: I think there are three small fixes:

1. Change the order of the deletes in `delete_pending_temp_tables`.
2. Delete the `moz_places_afterinsert_trigger`, since it's now redundant.
3. Change the expected frecency in the `api::matcher::tests::search` test from 1999 to 2000.

Reply:
ah yes, thank you! I actually started with a theory that the order of deletions was wrong, but then convinced myself I was confused about that.